Feature/accessibility

lakesare opened the pull request:
1. Made provider components accessible. Had to change some html&css along the way.
2. Made file-already-uploaded components accessible.
3. Made focus managed throughout Uppy in Dashboard. Now we should be able to navigate through Uppy entirely via the keyboard. I went through it with Mac's VoiceOver, everything's read well, although sometimes Informer gets interrupted by the shifted focus.

I thought about how to implement focus management for some time, and I tried a few options.
Basically if we tried to implement it via the `componentDidMount/Update`s with timeouts, focus may be fighting between multiple components, and sometimes it may be getting lost out of the current overlay.
What I did is I made it recalculate where focus should be every 250ms, and introduced 'superfocusable' elements - if they are rendered, focus will fall on them if it's not yet on some other superfocusable element.

Concerns:
- I didn't manage to check Dropbox integration, we'd need new credentials to check it.
- Is there a case where focus could be jumping back and forth between two element in a loop, do you see when this state may occur? (`dashboard/src/utils/createSuperFocus.js`)

Fixes #1415, fixes #1414, fixes #1413, fixes #1556, fixes #1351

arturi commented on packages/@uppy/dashboard/src/components/FileCard.js:
```diff
@@ -49,10 +42,8 @@ class FileCard extends Component {
-            if (i === 0) this.firstInput = el
-          }} /></fieldset>
+          onkeypress={this.tempStoreMetaOrSubmit} />
+      </fieldset>
```
Why is there no `data-uppy-super-focusable` here, will it focus on the first input?

arturi commented on packages/@uppy/url/src/UrlUI.js:
```diff
@@ -36,7 +29,8 @@ class UrlUI extends Component {
         onkeyup={this.handleKeyPress}
-        ref={(input) => { this.input = input }} />
+        ref={(input) => { this.input = input }}
+        data-uppy-super-focusable />
```
`ref={(input)` not needed here?

lakesare replied:
It's not, Dashboard will see `data-uppy-super-focusable` attribute rendered, and automatically focus on it.

arturi replied:
Ok, right, but I guess it was left here by mistake then? Could you remove?

lakesare replied:
@arturi, ah, it certainly was, I will! #TODO

lakesare replied:
It wasn't actually, turns out `this.input` is used in that component.

arturi replied:
Oh, ok, sorry about that then!

arturi commented on package.json:
```diff
@@ -147,5 +147,7 @@
+  "dependencies": {
+    "lodash.debounce": "^4.0.8"
+  }
 }
```
This should be in `@uppy/dashboard` dependencies probably?

lakesare replied:
@arturi, good point.
I also saw self-written debounce() function used in another plugin (I didn't change it for lodash's version, because I wasn't 100% sure if there are interchangeable).
In general I feel like that's some function we may want to use in other plugins.
What do you think?

arturi replied:
> What do you think, should we just be adding lodash.debounce to every plugin separately when we need it?

Yes, since they are installed separately like `npm i @uppy/core @uppy/dashboard` — can’t use root-level package.json for that.

I also saw self-written debounce() function used in another plugin (I didn't change it for lodash's version, because I wasn't 100% sure if there are interchangeable).

We should ask @goto-bus-stop if their debounce https://github.com/transloadit/uppy/blob/854654b135eece671288fadf52c08e97ff220c59/packages/%40uppy/core/src/Plugin.js#L7-L24 can be replaced with lodash? I guess there was a reason for a custom version maybe?

lakesare replied:
@arturi, understood, will move lodash.debounce to Dashboard's package.json #TODO.

goto-bus-stop replied:
My "debounce" only works for many synchronous calls. eg when a plugin calls setState() 10 times in a row synchronously it makes sure that the `updateUI` function is only called once at the end (by scheduling a ["microtask"](https://jakearchibald.com/2015/tasks-microtasks-queues-and-schedules/) using `Promise.resolve()`).
It doesn't support custom delays or anything that lodash has, so most likely lodash.debounce should be used instead.

arturi commented:
Thank you! 👍 Started to review gradually, will need to take another look and test still, but looking good so far!

arturi commented on packages/@uppy/dashboard/src/utils/createSuperFocus.test.js:
```diff
@@ -0,0 +1,10 @@
+
+describe('createSuperFocus', () => {
+  // superFocus.cancel() is used in dashboard
+  it('should return a function that can be cancelled', () => {
```
Confused, sorry, but where is `superFocus.cancel()` defined?

lakesare replied:
@arturi, it's defined on the function that's returned from `debounce()`!
(Thank you for starting to review this PR, I'll answer all comments when I get some more time)

arturi commented:
When I click on an image in Instagram or a file in Google Drive, it gets selected, the checkmark appears, and the state updates, and then the focus switches to the first `data-uppy-super-focusable` — the first item in the list. This causes the focus outline to flash and the view scrolls to the top. 

My first guess is we should call `.focus()` on an item after it’s selected? Because this only happens when selecting with mouse/click, when selecting with keyboard everything works as expected.

Adding `e.currentTarget.focus()` to `toggleCheckbox()` in `provider-views/index.js` seems to resolve the issue.

lakesare commented:
@arturi,
> When I click on an image in Instagram or a file in Google Drive, it gets selected, the checkmark appears, and the state updates, and then the focus switches to the first

I see that issue in safari, but not in chrome! Investigating.

lakesare commented:
It because safari thinks that after I click on the checkbox manually, document.activeElement becomes the body 🤔 
In fact safari sometimes thinks that usual buttons are not tabbable for some reason, I can't tab into the Google Drive, or other subfolders in breadcrumbs.

arturi commented on packages/@uppy/provider-views/src/style/uppy-ProviderBrowser-viewType--list.scss:
```diff
@@ -0,0 +1,63 @@
+
+    // Checkbox
+    > .uppy-ProviderBrowserItem-fakeCheckbox {
+      margin-right: 15px;
```
Is deep-deep nesting like this necessary here? Ideally I’d lift a level or so, it becomes a little hard to follow and too specific, I think.

lakesare replied:
@arturi, it isn't, it's just my personal preference, I think:
- it's easier to follow (because we can visualise the html hierarchy by reading css), and
- nesting is good (for the same reasons that we like to only have our Js variables accessible from a scope as little as possible).

But I understand it's an unusual preference, and in a library tight specificity may become a problem, I can unnest it.

arturi commented on packages/@uppy/dashboard/src/components/FileItem.js:
```diff
@@ -128,40 +128,51 @@ module.exports = function FileItem (props) {
           file={file}
           error={error}
+          isUploaded={isUploaded}
           {...props} />
```
Great, spent about an hour trying to figure out why the progress indicator remained a `<button>` on master, but turned into a `<div>` in your PR, if the PR is not touching the logic of that component 🙀 😂

arturi commented:
If Firefox, in provider-views, it sometimes focuses on the whole list of files, instead of the first item. Discussed this in a call, just noting here, so we don’t forget.

lakesare commented:
@arturi, does it look good to go now?

arturi commented:
Yes, thanks for addressing issues quickly! Will test again and merge after we release a patch with a hotfix.

arturi commented:
Merge conflict, unfortunately :( I think after this was merged: https://github.com/transloadit/uppy/pull/1549/files#diff-025cd2d037f664864c8d78edc6a2a85f

arturi commented:
Some other findings (could be a follow-up PR):

1. `uppy-DashboardFileCard-label` should wrap the input or have a `for` attribute, otherwise they are not linked, only placeholder text is read by VoiceOver.
2. Maybe we should add `role=alert` to `uppy-StatusBar-status` or `uppy-StatusBar-statusPrimary` (since `uppy-StatusBar-status` will then read every progress update, which might be too much?)

arturi commented on .eslintrc.json:
```diff
@@ -9,10 +9,20 @@
+    "jsdoc/check-types": ["error"],
+    "jsdoc/newline-after-description": ["error"],
+    "jsdoc/valid-types": ["error"]
   },
```
Cool! I wonder if this will result in many errors in other files that use JSDoc?

lakesare commented:
@arturi, > If that’s the case, then I guess you can safely accept your version. Are square [ and round ( brackets the same thing for that purpose? I think it’s better to use round ones for consistency, that’s what we do everywhere else.

I used it as an array, in order to have two elements without an excessive div (because preact doesn't do <React.Fragment>s yet).

arturi commented:
> I used it as an array, in order to have two elements without an excessive div (because preact doesn't do <React.Fragment>s yet).

Oh, ok, awesome, I didn’t know it works like that!

lakesare commented:
> Some other findings (could be a follow-up PR, just collecting here so we don’t forget)

I will look into these, but preferably in another PR, this one is so large it gives me merge-anxiety, I wish it was in master :-)

arturi commented:
As much as I’d like to merge this already, discovered another issue while testing :(

When you set `inline: true`, the focus from the page moves straight to the Dashboard. Which is likely not desirable, since you have other elements on the page which need to be in focus first, before you reach Dashboard for file selecting/uploading.

When in modal mode `inline: false` that’s a different thing — we want focus to always be in modal, until it’s closed.

Could we enable superfocus only when focus is inside the Dashboard? Then when we open modal, we could set focus inside it manually.

Evgenia added:

> Also when inline and someone wants to fill out a form close by, for example, focus will be moved to Uppy Dashboard on progress events

lakesare commented:
@arturi, thank you for noticing!
Pushed the less forceful behaviour for the inline mode, testing it now.

lakesare commented:
On brief check it works well in inline and modal modes, but please check too, my eyes may be glazing over it.

arturi commented on packages/@uppy/dashboard/src/index.js:
```diff
@@ -566,17 +579,35 @@ module.exports = class Dashboard extends Plugin {
+      //   BUT we have already, at least once, focused on uppy [this is done to avoid focusing related to the page initialisation]
+      if (isFocusInUppy || (isFocusOnBody && this.ifFocusedOnUppyAtLeastOnce)) {
+        this.superFocus(this.el, this.getPluginState().activeOverlayType)
+      }
```
Do we need to repeat the `isInformerPresent` logic for inline?

lakesare commented:
Checked inline&modal in Edge - good, but focus styles aren't visible anywhere (at least in browserstack),
checked inline&modal in Android Chrome (while typing too) - good,
checked inline&modal in Chrome, Firefox, Safari (while typing too) - good.

arturi commented:
Testing in IE11 — plugins icons require to tabs to switch, we need to add `focusable="false"` to svg icons in buttons https://www.sarasoueidan.com/blog/accessible-icon-buttons/.

(also a todo for later followup, just noting so that we don’t forget).

arturi commented:
More issues uncovered during testing :( :

1. When inline, during upload in the background (large file), click the add more files `+` — then slowly tab through all plugins, after Camera the focus falls to `uppy-StatusBar-actionCircleBtn` pause/resume button from underneath 😱 . Then it jumps back to “My Device”, so you are stuck, can’t get out of modal, can’t reach “back” button, unless you tab really quickly (faster than progress?).
2. When in modal, during upload in the background, add more files `+` — the focus on the first icon appears with delays (only after state is updated with upload progress?), same when you click “Instagram” — authenticate button is not in focus for some time, same goes for Link url input field.

lakesare commented:
Another issue for follow-up PR: uploaded image doesn't get focus in firefox.

lakesare commented:
My commit fixes 1st issue in chrome&firefox.
2nd issue is not present in chrome, but is present in firefox, trying to find out why it happens.

arturi commented on packages/@uppy/dashboard/src/index.js:
```diff
@@ -535,6 +536,15 @@ module.exports = class Dashboard extends Plugin {
+
+    if (this.opts.inline) {
+      document.addEventListener('keydown', this.handleKeyDownInInline)
+    }
```
Nitpicking: would `this.el.addEventListener` work?

lakesare replied:
_Kind of_.
With `this.el`, in Link overlay: tabbing from the page requires double tab.
With `document`, in Link overlay: tabbing from page MISSES the Submit button.

Changed it to `this.el` for now, but this can be considered a #TODO for feature ammendments.

arturi commented:
> started trapping focus in inline mode too

tested this, works well, much better now, since focus doesn’t leave overlay ever 👍 

But a little concerned about our moral rights 👨‍⚖ to trap user in our overlay when they are not really in modal. Basically with this change you can’t leave Uppy Dashboard unless you press on “done” or “back”.

lakesare commented:
I think overlays are enough of a modal. Person just clicked 'open something', they remember how to go back.
Either way it's the easy solution that shares code with the modal, I worry we may be losing focus in some cases if we try to do it via the `inert`.

lakesare commented:
#TODO upload button doesn't have focus (one file is uploading, and another is added) - small blue Upload (<button type="button" class="uppy-u-reset uppy-c-btn uppy-StatusBar-actionBtn" aria-label="Upload 1 file">Upload</button>).

lakesare commented:
> In firefox scroll jumps when one file is uploading, and we click ADD MORE.

This is a composite issue.

1. `el.focus({ preventDefault: true })` option does not work in firefox. It's possible to use a polyfill (https://gist.github.com/lakesare/b33cab4d7fcfeadbae7f532e46f47a50), but still makes the focus jump a little.
2. This happens because we use `throttle` instead of `debounce`. It means if updates are happening nonstop (because file is uploading), then we will focus on the animated ADD FILES before it's done animating, creating a jump.
3. The root of the issue is nonstop firefox state updates.

lakesare commented:
I added the library for `inert`.
There was no easy way to make the focus jump straight to current overlay (when some overlay is open).
Now we don't loop in the inline mode, now it's done how we initially wanted it.

`wicg-inert` uses MutationObserver` internally.
I added a `MutationObserver` polyfill, because tests were failing without it.
This polyfill needs a couple of polyfills of its own: https://github.com/WICG/inert/blob/master/karma.conf.js#L31.

Do you know if we need to inform end consumers of it, or does our babel handle it by any chance?

lakesare commented:
Locally `npm run test:endtoend:local` passes, in both Firefox and Chrome.
Dev build works in Saucelabs Chrome, works in Browserstack IE, and local Firefox.
I'm not sure why the test is failing.

lakesare commented:
I reverted changed that introduced the `inert` html property.
It would require too many polyfills, and it can introduce performance issues.
We decided to delay this solution until at least some browsers implement `inert` natively.

lakesare commented:
A list of unresolved TODOs for the **follow-up PR**:

1. `uppy-DashboardFileCard-label` should wrap the input or have a `for` attribute, otherwise they are not linked, only placeholder text is read by VoiceOver.
2. Maybe we should add `role=alert` to `uppy-StatusBar-status` or `uppy-StatusBar-statusPrimary` (since `uppy-StatusBar-status` will then read every progress update, which might be too much?)
3. Testing in IE11 — plugin selection buttons require the tab button to be pressed 2 times to switch to the next button, we need to add `focusable="false"` to svg icons in buttons https://www.sarasoueidan.com/blog/accessible-icon-buttons/. - 
4. Another issue for follow-up PR: uploaded image doesn't get focus in firefox.
5. Firefox's default outlines are not visible.
6. Upload button doesn't have focus styles (one file is uploading, and another is added) - small blue Upload (Upload).

lakesare commented:
Edge:
modal, inline - works

IE11:
modal - works
inline - works in general, but focus gets lost sometimes (when overlay closes e.g.)

arturi commented:
Thank you for the quick turnaround again!

>  1. We focus when isFocusNowhere, because when the element we were focused on disappears (e.g. an overlay), - focus gets lost. If user is typing something somewhere else on the page, - focus won't be 'nowhere'.

Thought about that some more and tested: when I click on the document “nowhere”, focus is set on the body, but Uppy then steals it back. Same when you are tabbing through everything and you then move to the browser window, address bar, then you are supposed to go to body or “nowhere”, Uppy immediately steals focus.

Not saying this is a major blocked, but sharing a concern.

lakesare commented:
@arturi, I see what you mean. In Safari for example we can't simultaneously press checkbox with SPACE multiple times, and have some large file be downloaded in the meantime.

I think we should ask ourselves whether we'd stumble upon worse issues if we didn't use superfocus, what do you think?

What I tried:
1. Introducing a second argument to the uppy.setState: `setState(patch, typeOfUpdate)`.
Then I set `typeOfUpdate` to 'upload-progress' in every relevant place. Issue is, there are a lot of relevant places, do we really want to remember setting this second argument every time?

2. Removing `(isFocusNowhere && this.ifFocusedOnUppyAtLeastOnce)` altogether. This makes focus get lost whenever some element disappears (e.g. there was a pause button, but there is none now, because upload is complete, e.g. when google-drive folder is selected in safari).

3. ?? Thinking.

lakesare commented:
Fixed it. Checked IE11 inline, Firefox inline&modal, Android inline, Chrome inline&modal, Safari inline.

arturi commented:
Sweet! ✨ Works much better now. I think checking if prev focus was on Uppy, instead of if it was ever on Uppy, is a very good idea 👍

In my tests superfocus still does sometimes steal focus back to Uppy when you try to click an input while upload is in progress / state is updating.

It is fixed for me by adding an additional `this.superFocus.cancel()` to `recordIfFocusedOnUppyRecently`, so it can cancel the debounced superfocus when focus moves away from Uppy (if I understand correctly).

Committed the fix, hope that’s ok, please try without the fix (easy way to test, btw, is `setInterval(() => uppy.setState({123: '123'}), 500)` in console) and feel free to move/revert.

arturi commented:
Found another bug for future PR (unrelated to this, since it happens on master now):

1. In console, do `setInterval(() => uppy.setState({123: '123'}), 500)`.
2. Try to edit filter input in Google Drive — it’s buggy, loosing keystrokes, restores deleted text, drives users nuts.

#TODO

lakesare commented:
Oh right! Good catch with `this.superFocus.cancel()`.

Fixed the filter input in Google Drive while I was at it.

Found the issue with setInterval(500) and empty Google Drive folders (focus would be jumping to the first element all the time), - fixed it.

arturi commented:
Thank you for so much hard work and research on this!

lakesare commented:
@arturi, hopefully it goes well!
